File: level1/lib/detection/analysis_dispatcher.dart
import 'dart:async';

/// Serializes [process] invocations behind a single latest-wins pending slot.
///
/// At most one [process] call is in flight at a time. Adding an input while the
/// worker is busy replaces the pending slot, dropping the previous pending
/// input. Errors from [process] are caught and routed to [onError]; the worker
/// loop always continues.
class AnalysisDispatcher<T> {
  AnalysisDispatcher({required this.process, this.onError});

  final Future<void> Function(T input) process;
  final void Function(Object error, StackTrace stackTrace)? onError;

  Object? _pending;
  bool _hasPending = false;
  bool _processing = false;
  bool _disposed = false;

  void add(T input) {
    if (_disposed) return;
    _pending = input;
    _hasPending = true;
    if (!_processing) {
      _processing = true;
      _drain();
    }
  }

  void _drain() {
    if (_disposed || !_hasPending) {
      _processing = false;
      return;
    }
    final input = _pending! as T;
    _hasPending = false;
    Future.sync(() => process(input)).then((_) {
      if (_disposed) {
        _processing = false;
        return;
      }
      _drain();
    }, onError: (Object error, StackTrace stackTrace) {
      onError?.call(error, stackTrace);
      if (_disposed) {
        _processing = false;
        return;
      }
      _drain();
    });
  }

  /// Clears the pending slot and stops the loop once in-flight work completes.
  Future<void> dispose() async {
    _disposed = true;
    _hasPending = false;
    _pending = null;
    while (_processing) {
      await Future<void>.delayed(Duration.zero);
    }
  }
}

File: level2/android/app/src/main/kotlin/io/securitycam/level2/detection/pipeline/AnalysisDispatcher.kt
package io.securitycam.level2.detection.pipeline

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

/**
 * Serializes [process] invocations behind a single latest-wins pending slot
 * (port of `lib/detection/analysis_dispatcher.dart`). At most one [process]
 * call is in flight at a time; adding an input while busy replaces the pending
 * slot. Errors are routed to [onError] and the loop always continues.
 */
class AnalysisDispatcher<T>(
    private val scope: CoroutineScope,
    private val process: suspend (T) -> Unit,
    private val onError: (Throwable) -> Unit = {},
) {
    private val lock = Any()
    private var pending: T? = null
    private var hasPending = false
    private var processing = false
    private var drainJob: Job? = null

    @Volatile
    private var disposed = false

    fun add(input: T) {
        synchronized(lock) {
            if (disposed) return
            pending = input
            hasPending = true
            if (!processing) {
                processing = true
                drainJob = scope.launch { drain() }
            }
        }
    }

    private suspend fun drain() {
        while (true) {
            val input: T
            synchronized(lock) {
                if (disposed || !hasPending) {
                    processing = false
                    drainJob = null
                    return
                }
                @Suppress("UNCHECKED_CAST")
                input = pending as T
                hasPending = false
            }
            try {
                process(input)
            } catch (t: Throwable) {
                onError(t)
            }
        }
    }

    /** Clears the pending slot and stops the loop once in-flight work completes. */
    suspend fun dispose() {
        val job: Job?
        synchronized(lock) {
            disposed = true
            hasPending = false
            pending = null
            job = drainJob
        }
        job?.join()
    }
}